pkg/build/controller/controller.go

```diff
@@ -108,6 +163,44 @@ func (bc *BuildController) HandleBuild(build *buildapi.Build) error {
 	return nil
 }
 
+func (bc *BuildController) nextBuild(namespace, buildConfigName string) (*buildapi.Build, error) {
+	hasRunningBuilds := false
+	builds, err := buildutil.BuildConfigBuilds(bc.OSClient.Builds(namespace), buildConfigName, func(b buildapi.Build) bool {
+		// Detect if there are running builds
+		switch b.Status.Phase {
+		case buildapi.BuildPhaseRunning, buildapi.BuildPhasePending:
+			hasRunningBuilds = true
+			return false
+		}
+		// Only 'new' build can be scheduled to run next
+		return b.Status.Phase == buildapi.BuildPhaseNew
+	})
+	if hasRunningBuilds || err != nil {
+		return nil, err
+	}
+	var (
+		nextBuild           *buildapi.Build
+		previousBuildNumber int64
+	)
+	for i, b := range builds.Items {
+		buildNumber, err := buildutil.BuildNumber(&b)
+		if err != nil {
+			glog.Errorf("Error getting build number: %v", err)
+			continue
+		}
+		if previousBuildNumber == 0 {
+			nextBuild = &builds.Items[i]
+			previousBuildNumber = buildNumber
+			continue
+		}
```

Comment:
Put the strategy behind an interface.

On Mon, Apr 18, 2016 at 6:43 AM, Michal Fojtik <notifications@github.com>
wrote:

> In pkg/build/controller/controller.go
> <https://github.com/openshift/origin/pull/8453#discussion_r60037780>:
>
> > +	}
> > +	var (
> > +		nextBuild           *buildapi.Build
> > +		previousBuildNumber int64
> > +	)
> > +	for i, b := range builds.Items {
> > +		buildNumber, err := buildutil.BuildNumber(&b)
> > +		if err != nil {
> > +			glog.Errorf("Error getting build number: %v", err)
> > +			continue
> > +		}
> > +		if previousBuildNumber == 0 {
> > +			nextBuild = &builds.Items[i]
> > +			previousBuildNumber = buildNumber
> > +			continue
> > +		}
>
> @bparees <https://github.com/bparees> one concern with having three
> policies I have is that the switching of policies on-the-fly can cause a
> mess :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8453/files/68d0bdac5916b17673695e06451439bfa55941b9#r60037780>
>